TRUNK-6160 Added diagnoses support to encounter service

icrc-loliveira opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111: Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber: JiraIssueTitle' -->
## Description of what I changed
Added diagnoses support to encounter service.
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first! -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it! -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-6160

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -425,6 +428,33 @@ public void saveEncounter_shouldCascadeSaveToContainedConditions() {
+		diagnosis2.setCertainty(ConditionVerificationStatus.PROVISIONAL);
+		
+		// replay
+		Set<Diagnosis> diagnoses = new HashSet<>();
```
What are you using this hashset for?

icrc-loliveira replied:
Nothing, some leftovers from an earlier code.

dkayiwa commented:
@icrc-loliveira your tests do not demonstrate the reported problem because they all pass without a single change in `EncounterServiceImpl`

tendomart commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -1047,6 +1093,23 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
+	@Test
+	public void unvoidEncounter_shouldCascadeUnvoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
```
Redundant space may be removed

icrc-loliveira commented:
Hello @dkayiwa,
I have fixed the test to create diagnoses. (I based my test on the existing ones, so is possible that those are also not validating as expected.)

For the void and unvoided functions. It seems, based on the tests, that they are already taking into account the Diagnoses. So I removed my changes but leave the tests. (I have no idea how this is being handled but if it is working that is good enougt for me :) )

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -447,10 +456,7 @@ public Encounter unvoidEncounter(Encounter encounter) throws APIException {
 		}
 		
-		String voidReason = encounter.getVoidReason();
-		if (voidReason == null) {
```
Do we really need this change?

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -974,7 +1001,23 @@ public void voidEncounter_shouldCascadeToOrders() {
+		encounterService.voidEncounter(encounter, "Just Testing");
 
+		Diagnosis diagnosis = Context.getDiagnosisService().getDiagnosis(1); 
+		assertTrue(diagnosis.getVoided());
```
Can you assert that this was not voided before.

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -1048,6 +1091,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
+	public void unvoidEncounter_shouldCascadeUnvoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a voided encounter that has some voided obs
```
Voided obs or voided diagnosis?

icrc-loliveira replied:
Done

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -1048,6 +1091,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
+		encounterService.unvoidEncounter(encounter);
+
+		Diagnosis diagnosis = Context.getDiagnosisService().getDiagnosis(2);
+		assertFalse(diagnosis.getVoided());
```
Can you assert that this was voided before.

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -974,7 +1001,23 @@ public void voidEncounter_shouldCascadeToOrders() {
+	public void voidEncounter_shouldCascadeVoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a nonvoided encounter that has some obs
```
That has some obs or diagnosis?

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -1048,6 +1099,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
+		Encounter encounter = encounterService.getEncounter(2);
+		encounterService.unvoidEncounter(encounter);
+
+		Diagnosis diagnosis = Context.getDiagnosisService().getDiagnosis(2);
```
Can you assert that the diagnosis was voided before the un-voiding of the encounter?

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -1048,6 +1099,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a voided encounter that has some voided Diagnoses
+		Encounter encounter = encounterService.getEncounter(2);
```
Can you also assert that the encounter is voided?

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -974,7 +1001,31 @@ public void voidEncounter_shouldCascadeToOrders() {
+		// Test that diagnoses is unvoided
+		Diagnosis unvoidedDiagnosis = encounter.getDiagnoses().iterator().next();
+		assertEquals(unvoidedDiagnosis.getDiagnosisId(), 1);
+		assertTrue(!unvoidedDiagnosis.getVoided());
```
How about using assertFalse?

dkayiwa commented on api/src/test/java/org/openmrs/api/EncounterServiceTest.java:
```diff
@@ -974,7 +1001,31 @@ public void voidEncounter_shouldCascadeToOrders() {
+		Diagnosis unvoidedDiagnosis = encounter.getDiagnoses().iterator().next();
+		assertEquals(unvoidedDiagnosis.getDiagnosisId(), 1);
+		assertTrue(!unvoidedDiagnosis.getVoided());
+		
```
Can you also assert that the encounter is not voided?
